cpp/src/arrow/io/buffered.cc

```diff
@@ -178,5 +217,215 @@ Status BufferedOutputStream::Flush() { return impl_->Flush(); }
+
+  int64_t buffer_size() const { return buffer_size_; }
+
+  std::shared_ptr<InputStream> Detach() {
+    std::lock_guard<std::mutex> guard(lock_);
+    is_open_ = false;
+    return std::move(raw_);
+  }
+
+  void RewindBuffer() {
+    // Invalidate buffered data, as with a Seek or large Read
+    buffer_pos_ = bytes_buffered_ = 0;
+  }
+
+  Status BufferIfNeeded() {
+    // TODO more graceful handling of zero-copy sources.
+    if (bytes_buffered_ == 0) {
+      // Fill buffer
+      if (!buffer_ || buffer_.use_count() > 1) {
+        // A reference to the internal buffer has been exported, so we allocate
+        // a new buffer here. This optimization avoids unnecessary memory
+        // allocations and copies, while making exported buffers safe for use
+        // if the caller does not destroy them
+        buffer_ = nullptr;
+        RETURN_NOT_OK(ResetBuffer());
+      }
+      RETURN_NOT_OK(raw_->Read(buffer_size_, &bytes_buffered_, writable_buffer_data_));
+      buffer_pos_ = 0;
+
+      // Do not make assumptions about the raw stream position
+      raw_pos_ = -1;
+    }
+    return Status::OK();
+  }
+
+  void ConsumeBuffer(int64_t nbytes) {
+    buffer_pos_ += nbytes;
+    bytes_buffered_ -= nbytes;
+  }
+
+  Status Read(int64_t nbytes, int64_t* bytes_read, void* out) {
+    std::lock_guard<std::mutex> guard(lock_);
+    DCHECK_GT(nbytes, 0);
+
+    if (nbytes < buffer_size_) {
+      // Pre-buffer for small reads
+      RETURN_NOT_OK(BufferIfNeeded());
+    }
+
+    if (nbytes > bytes_buffered_) {
+      // Copy buffered bytes into out, then read rest
+      memcpy(out, buffer_data_ + buffer_pos_, bytes_buffered_);
+      RETURN_NOT_OK(raw_->Read(nbytes - bytes_buffered_, bytes_read,
+                               reinterpret_cast<uint8_t*>(out) + bytes_buffered_));
+      // Do not make assumptions about the raw stream position
+      raw_pos_ = -1;
+      *bytes_read += bytes_buffered_;
+      RewindBuffer();
+    } else {
+      memcpy(out, buffer_data_ + buffer_pos_, nbytes);
+      *bytes_read = nbytes;
+      ConsumeBuffer(nbytes);
+    }
+    return Status::OK();
+  }
+
+  Status Read(int64_t nbytes, std::shared_ptr<Buffer>* out) {
+    if (bytes_buffered_ == 0 && supports_zero_copy_) {
+      raw_pos_ = -1;
+      return raw_->Read(nbytes, out);
+    }
+
+    if (nbytes > bytes_buffered_) {
+      // Cannot do zero copy read, instead allocate buffer and read into that
+      std::shared_ptr<ResizableBuffer> buffer;
+      RETURN_NOT_OK(AllocateResizableBuffer(pool_, nbytes, &buffer));
+
+      int64_t bytes_read = 0;
+      RETURN_NOT_OK(Read(nbytes, &bytes_read, buffer->mutable_data()));
+      if (bytes_read < nbytes) {
+        RETURN_NOT_OK(buffer->Resize(bytes_read));
+        buffer->ZeroPadding();
+      }
+      *out = buffer;
+    } else {
+      RETURN_NOT_OK(BufferIfNeeded());
+
+      // Slice the internal buffer. On subsequent reads, a new buffer will be
```

Comment:
This may actually be wasteful. In the real world, you can have a 4-byte slice of the underlying 8192-byte buffer (or perhaps even larger, if you want to hide latencies to a networked file system), and it will keep alive the whole thing. I think it's more robust to do the conservative thing, i.e. always allocate a new buffer and copy to it.

If some performance profile shows that allocation is a significant performance hog, we may investigate a caching memory pool (though with jemalloc and other modern allocators, I'm not sure if that's bound to happen).

Reply:
It's a bit unsatisfying to copy twice if we don't have to -- e.g. a use case here will be reading data pages (which are fairly small) from Parquet files with a fairly large (1MB or more) read buffer. We've already had someone note issues with heavy allocation activity on this hot path (I'd have to dig up the JIRA / mailing list thread about it).

I can remove the zero copy optimizations for now; we can perhaps address the issue in another way. For example, I am interested in creating a "buffer pool" (maybe this is what you are suggesting also) that allows buffers to be created and then returned to the pool. This could be made opaque to a class like `BufferedInputStream` via an abstract BufferAllocator, see:

https://issues.apache.org/jira/browse/ARROW-1470

This way, by passing the right implementation of `BufferAllocator`, you could avoid allocations on the hot path (but you would still have copies).

Another option is that we could have a special method on `BufferedInputStream` that permits this buffer-exporting behavior, so that the consumer code must explicitly opt into it, so they are aware that they could be causing more memory to be allocated if they are retaining non-copies references to the internal buffer.